examples/regression/4-tutorials/plot_main-tutorial-regression.py

```diff
@@ -0,0 +1,795 @@
+        X_train.ravel(),
+        y_train.ravel(),
+        X_test.ravel(),
+        y_mesh.ravel(),
+        (1.96*noise*X_test).ravel(),
+        y_pred[strategy].ravel(),
+        y_pis[strategy][:, 0, 0].ravel(),
+        y_pis[strategy][:, 1, 0].ravel(),
+        ax=coord,
+        title=strategy
+    )
+
+
+##############################################################################
+# We can observe that all of the strategies except CQR seem to have similar
+# constant prediction intervals.
+# On the other hand, the CQR strategy offers a solution that adapts the
+# prediction intervals to the local noise.
+
+fig, ax = plt.subplots(1, 1, figsize=(7, 5))
+ax.plot(X_test, 1.96*2*noise*X_test, ls="--", color="k", label="True width")
+for strategy in STRATEGIES:
+    ax.plot(
+        X_test,
+        y_pis[strategy][:, 1, 0] - y_pis[strategy][:, 0, 0],
+        label=strategy
+    )
+ax.set_xlabel("x")
+ax.set_ylabel("Prediction Interval Width")
+_ = ax.legend(fontsize=10, loc=[1, 0.4])
+
+
+##############################################################################
+# One can observe that all the strategies behave in a similar way as in the
+# first example shown previously. One exception is the CQR method which takes
+# into account the heteroscedasticity of the data. In this method we observe
+# very low interval widths at low values of :math:`x`.
+# This is the only method that
+# even slightly follows the true width, and therefore is the preferred method
+# for heteroscedastic data. Notice also that the true width is greater (lower)
+# than the predicted width from the other methods at :math:`x \gtrapprox 3``
+# (:math:`x \leq 3`). This means that while the marginal coverage correct for
+# these methods, the conditional coverage is likely not guaranteed as we will
+# observe in the next figure.
+
+def get_heteroscedastic_coverage(y_test, y_pis, STRATEGIES, bins):
+    recap = {}
+    for i in range(len(bins)-1):
+        bin1, bin2 = bins[i], bins[i+1]
+        name = f"[{bin1}, {bin2}]"
+        recap[name] = []
+        for strategy in STRATEGIES:
+            indices = np.where((X_test >= bins[i]) * (X_test <= bins[i+1]))
+            y_test_trunc = np.take(y_test, indices)
+            y_low_ = np.take(y_pis[strategy][:, 0, 0], indices)
+            y_high_ = np.take(y_pis[strategy][:, 1, 0], indices)
+            score_coverage = regression_coverage_score(
+                y_test_trunc[0], y_low_[0], y_high_[0]
+            )
+            recap[name].append(score_coverage)
+    recap_df = pd.DataFrame(recap, index=STRATEGIES)
+    return recap_df
+
+
+bins = [0, 1, 2, 3, 4, 5]
+heteroscedastic_coverage = get_heteroscedastic_coverage(
+    y_test, y_pis, STRATEGIES, bins
+)
+
+fig = plt.figure()
+heteroscedastic_coverage.T.plot.bar(figsize=(12, 4), alpha=0.7)
+plt.axhline(0.95, ls="--", color="k")
+plt.ylabel("Conditional coverage")
+plt.xlabel("x bins")
+plt.xticks(rotation=0)
+plt.ylim(0.8, 1.0)
+plt.legend(loc=[1, 0])
+
```

Comment:
add a plt.show() to remove the log from matplotlib

Reply:
Done